ampd/src/prometheus_metrics/monitor.rs

```diff
@@ -0,0 +1,329 @@
+
+    fn start_metrics_processing(
+        server: Metrics,
+        mut metrics_rx: mpsc::Receiver<MetricsMsg>,
+        cancel: CancellationToken,
+    ) {
+        tokio::spawn(async move {
+            loop {
+                tokio::select! {
+
+                    msg = metrics_rx.recv() => {
+                        match msg {
+                            Some(msg) => {
+                                if let Err(e) = server.handle_message(msg) {
+                                    tracing::error!("Failed to handle metrics message: {:?}", e);
+                                }
+                            }
+                            None => {
+                                break;
+                            }
+                        }
+                    }
+                    _ = cancel.cancelled() => {
+                        break;
+                    }
+                }
+            }
+        });
+    }
+}
+
+async fn status() -> (StatusCode, Json<Status>) {
+    (StatusCode::OK, Json(Status { ok: true }))
+}
+
+async fn gather_metrics(registry: &Registry) -> (StatusCode, String) {
+    match crate::prometheus_metrics::metrics::gather(registry) {
+        Ok(metrics) => (StatusCode::OK, metrics),
+        Err(e) => (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            format!("Failed to gather metrics: {}", e),
+        ),
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct Status {
+    ok: bool,
+}
+
+#[cfg(test)]
+pub mod tests {
+    use std::net::{SocketAddr, TcpListener};
+    use std::time::Duration;
+
+    use tokio::test as async_test;
+
+    use super::*;
+
+    fn test_bind_addr() -> Option<SocketAddrV4> {
+        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
+
+        match listener.local_addr().unwrap() {
+            SocketAddr::V4(addr) => Some(addr),
+            SocketAddr::V6(_) => panic!("unexpected address"),
+        }
+    }
+
+    pub fn test_server_setup() -> (
+        Option<SocketAddrV4>,
+        Server,
+        MetricsClient,
+        CancellationToken,
+    ) {
+        let bind_address = test_bind_addr();
+        let (server, metrics_client) = Server::new(bind_address).expect("Failed to create server");
+        let cancel = CancellationToken::new();
+
+        (bind_address, server, metrics_client, cancel)
+    }
+
+    pub fn test_dummy_server_setup() -> (Server, MetricsClient, CancellationToken) {
+        let (server, metrics_client) = Server::new(None).expect("Failed to create server");
+        let cancel = CancellationToken::new();
+
+        (server, metrics_client, cancel)
+    }
+
+    fn send_metrics_client_msg(
+        metrics_client: &MetricsClient,
+        msg: MetricsMsg,
+        number_of_messages: usize,
+    ) {
+        for i in 0..number_of_messages {
+            metrics_client
+                .send_metrics_msg(msg.clone())
+                .unwrap_or_else(|_| panic!("Failed to send message {}", i));
+        }
+    }
+
+    #[async_test]
```

Comment:
```suggestion
    #[async_test, start_paused = true]
```

If we can avoid it, we don't want actual wait time in tests, because that bloats the overall time it takes to execute them. with `start_paused` tokio will actually artificially increase the clock time whenever the thread is idle, so sleeps just jump immediately to when they are done